docs: Improve getting started (#1095)
vercel/turborepo: docs/pages/docs/core-concepts/workspaces.mdx

---
title: Workspaces
description: Use workspaces to organize your turborepo apps and packages.
---

import Callout from "../../../components/callout";
import HeartIcon from "@heroicons/react/solid/HeartIcon";
import { Tabs, Tab } from '../../../components/Tabs'

# Getting started with workspaces

Workspaces are a feature implemented by package managers to assist with working on multiple apps and packages in the same codebase.

Turborepo is compatible with three package managers, [npm](https://docs.npmjs.com/cli/v8/using-npm/workspaces#description), [Yarn](https://classic.yarnpkg.com/lang/en/docs/workspaces/), and [pnpm](https://pnpm.io/workspaces), each with its own implementation of workspaces and functionality. Your package manager will determine how your apps and packages are organized in your workspaces and run in your turborepo.

## Configuring workspaces

> When using workspaces, you'll first need to define the location of your workspace, or where in the file structure your package manager should look for your workspace.

Author: ```suggestion
To use workspaces, you'll first need to declare their file system locations to your package manager
```

> A common pattern is to have an `apps` folder and a `packages` folder. The `apps` folder can contain workspaces for launchable apps, such as a `next.js` app. The `packages` folder can contain workspaces for packages that are used by either an app or another package.

Author: ```suggestion
A convention is to have an `apps` folder and a `packages` folder. The `apps` folder can contain workspaces for launchable apps, such as a Next.js or Svelte app. The `packages` folder can contain workspaces for packages that are used by either an app, or another package.
```

> You should consider how to name your workspace in a monorepo, since this will be different than working in single repositories. To avoid npm naming collisions, you can use the name of the monorepo and name of the workspace, separated by a slash and preceded by an `@`. For example, `@todolist/api`.

Author: ```suggestion
<Callout>
Unlike when working in a single-package repository, you'll need to consider how to name your workspaces. If you avoid npm naming collisions, you can use an npm organization or user scope to namespace your packages. For example, you might set the `name` key in `./apps/api/package.json` to `@mycompany/api`.
</Callout>
```

    ```json
    {
      "name": "monorepo",
      "version": "1.0.0",
      "workspaces": [
        "docs",
        "apps/*",
        "packages/*"
      ]
    }
    ```

  </Tab>

  <Tab>

    Add the folders you want to configure as workspaces to the `pnpm-workspace.yaml` file that exists in your root directory. This file contains a list of workspace folders in the form of globs:

    ```yaml
    packages:
      - "docs"
      - "apps/*"
      - "packages/*"
    ```

  </Tab>
</Tabs>

Each folder that matches any of the globs in the list is considered a workspace folder. In the examples, the `apps` and `packages` folders contain folders that are workspaces, and the `docs` folder itself is a workspace.

```
monorepo/
├─ docs/
├─ apps/
│  ├─ api/
│  ├─ mobile/
├─ packages/
│  ├─ tsconfig/
│  ├─ config/
├─ sdk/
```

Using the same workspace configuration, the `monorepo/docs/`, `monorepo/apps/api/`, `monorepo/apps/mobile/`, `monorepo/packages/config/`, and `monorepo/packages/tsconfig/` folders are all considered workspaces.
The `monorepo/sdk/` folder is not considered a workspace and it is not included in the workspace configuration.

## Managing workspaces

> When you move, delete, or rename your workspaces, you will have to make sure that all folders linked within your `package.json` matches. Anytime you change the configuration of your workspace, make sure you all dependencies of the workspace are also updated. Re-run your npm client's install command to check your configuration. If there are any problems after that, you may have to delete your `node_modules` folder and run an install again.

Reviewer: ```suggestion
When you move, delete, or rename your workspaces, you will have to make sure that all folders linked within your `package.json` matches. Anytime you change the configuration of your workspace, make sure all the dependencies of the workspace are also updated. Re-run your npm client's install command to check your configuration. If there are any problems after that, you may have to delete your `node_modules` folder and run an install again.
```

## Managing dependencies

> To manage dependencies within workspaces in your monorepo, you will need to run commands that will only manage the dependencies of each workspace rather than for entire monorepo.

Reviewer: ```suggestion
To manage dependencies within workspaces in your monorepo, you will need to run commands that will only manage the dependencies of each workspace rather than for the entire monorepo.
```

---
title: Workspaces
description: Use workspaces to organize your turborepo apps and packages.
---

import Callout from "../../../components/callout";
import HeartIcon from "@heroicons/react/solid/HeartIcon";
import { Tabs, Tab } from '../../../components/Tabs'

# Getting started with workspaces

Workspaces are a feature implemented by package managers to assist with working on multiple apps and packages in the same codebase.

Turborepo is compatible with three package managers, [npm](https://docs.npmjs.com/cli/v8/using-npm/workspaces#description), [Yarn](https://classic.yarnpkg.com/lang/en/docs/workspaces/), and [pnpm](https://pnpm.io/workspaces), each with its own implementation of workspaces and functionality. Your package manager will determine how your apps and packages are organized in your workspaces and run in your turborepo.

## Configuring workspaces

To use workspaces, you'll first need to declare their file system locations to your package manager

> A convention is to have an `apps` folder and a `packages` folder. The `apps` folder can contain workspaces for launchable apps, such as a Next.js or Svelte app. The `packages` folder can contain workspaces for packages that are used by either an app, or another package.

Reviewer 2: "A convention" sounds a bit odd.

Can we change this to "A common convention" or perhaps" The recommended convention"?

Author: ```suggestion
A common convention is to have an `apps` folder and a `packages` folder. The `apps` folder can contain workspaces for launchable apps, such as a Next.js or Svelte app. The `packages` folder can contain workspaces for packages that are used by either an app, or another package. This isn't a requirement, just a suggestion we've found useful.
```
